src/core/thread/network_data_notifier.cpp

```diff
@@ -152,6 +166,38 @@ void Notifier::HandleCoapResponse(Error aResult)
     }
 }
 
+#if OPENTHREAD_FTD && OPENTHREAD_CONFIG_BORDER_ROUTER_ENABLE && OPENTHREAD_CONFIG_BORDER_ROUTER_REQUEST_ROUTER_ROLE
+void Notifier::RequestRouterRoleIfEligible(void)
+{
+    VerifyOrExit(Get<Mle::MleRouter>().IsChild() && Get<Mle::MleRouter>().IsRouterEligible());
+
+    // RouterUpgradeThreshold can be explicitly set to zero in some of
+    // cert tests to disallow device to become router.
+
+    VerifyOrExit(Get<Mle::MleRouter>().GetRouterUpgradeThreshold() != 0);
+
+    // Check the local network data to ensure that we are a border
+    // router providing IP connectivity and therefore eligible to
+    // request router role upgrade with `kBorderRouterRequst`
+    // status.
+
+    VerifyOrExit(Get<Local>().CountBorderRouters(kAnyRole) > 0);
+
+    if (Get<RouterTable>().GetActiveRouterCount() >= Get<Mle::MleRouter>().GetRouterUpgradeThreshold())
```

Comment:
Here it should probably also check that ActiveRouterCount < MAX_ROUTERS ?

Reply:
I believe that check would be implicitly covered by the `VerifyOrExit()` line immediately below.

`Mle::kRouterUpgradeBorderRouterRequestThreshold` is currently set to `2`.

We could add a `static_assert` to ensure that `Mle::kRouterUpgradeBorderRouterRequestThreshold` is less than `Mle::kMaxRouters`.

Reply:
But the CountBorderRouters(kRouterRoleOnly) function immediately below doesn't check anything about the number of active routers? Only counts devices providing network data that 'looks like a BR'.  So in a situation that there are 32 allocated Routers already, the BR device in REED role would request ADDR_SOL.req every time some network data change occurs. (If I read the code correctly.)

Adding the static_assert is a good idea.

Reply:
Ah, yes. There should be a check against `kMaxRouters` somewhere.

Reply:
Yes. Good point. We need to exit if we are already at max routers (I guess it may be harmless to request but wasteful for sure). Will update it in the next push